fix #339: Send mouse events to the original object when mouse dragging

jestor opened the pull request:
- Prevents window Focus changes on mouse release
- MouseOut is still sent on mouse release

I'm not sure about the ramifications of the new logic in `mouseClicked`. I've done a fair amount of testing against fyne_demo and can't see any new weird behaviour or regressions.

The MouseUp/Down events are driven by `window.runEventQueue` unlike MouseIn/Out so the unittests use a 200 millisecond timeout to wait for them to trigger. A better approach might be for unittesting to force event queue processing.

andydotxyz commented:
The test for this - adding to a WaitGroup in a test, but Done() within a callback of a separate object seems likely to trip us up in the future.
You note that different mouse event types are handled differently - what if they all went through the event queue then we could have a private method like eventQueueEmpty()? Would that allow this test to read much better?

jestor commented:
> The test for this - adding to a WaitGroup in a test, but Done() within a callback of a separate object seems likely to trip us up in the future.
> You note that different mouse event types are handled differently - what if they all went through the event queue then we could have a private method like eventQueueEmpty()? Would that allow this test to read much better?

Yeah - I'll see if that lets me remove all the sync and time based logic in the unittests.

jestor commented:
I've added MouseIn/Out into the event queue system and modified the unittests as necessary. It adds quite a bit of bloat: `w.waitForEvents()` after every mouse operation.
In hindsight I could have left MouseIn/Out out of the queue as my draggable test is only watching MouseUp/Down... so if you want: I could revert a lot of that and still make it robust.

andydotxyz commented:
Looking at this it seems correct.
That is a lot of "bloat" I suppose - but it is ensuring accurate behaviour is it not?

jestor commented:
Yeah, I think that MouseIn/Out should be driven by the event queue. Which in turn requires us to perform a wait on the queue. If you're happy with it then I am.
